### PR TITLE
Add metrics for config reloads and config hash

### LINKED ISSUE #2040: Add metrics for config reloads
We need to add the following metrics to help debug issues with config reloads:

`agent_config_last_reload_success_timestamp_seconds`
`agent_config_last_reload_successful`
`agent_config_hash`

The first two exist in Prometheus and the third exists in Alertmanager and the Loki, Tempo, Mimir ecosystem.

### PR DESCRIPTION
**Pull request by jcreixell**
#### PR Description

- Add metrics for config reloads
  - Adds gauges to help debug issues with config reloads
- Add metric reporting config hash
  - Adds a gauge reporting a sha256 hash of the config (before
    expansion) via label. This should be useful to debug config reload
    issues.
  - The gauge is implemented for agent, agentctl and agent_flow and
    supports static, dynamic and remote config.
 
#### Which issue(s) this PR fixes

Fixes #2040 

#### Notes to the Reviewer

- I think I instrumented the code in the right places but might have missed something, please double check
- ~I had to namespace the metric name for flow as I would get a double registration otherwise. Amy ideas on how to do this better without too much complexity?~
- Is there a risk of cardinality explosion for the hash metric? (for example with dynamic configs)
- I assume we don't test internal metrics
- I assume we don't document these metrics, please correct if I am wrong
- Anything else I missed?

#### PR Checklist

- [x]  CHANGELOG updated

### CODE REVIEW

File: cmd/agent/entrypoint.go
Comment (mattdurham) [inline cmd/agent/entrypoint.go:72]: Check with Goutham but I wonder if its better to have success/failure be a label so we can track the two separately, it does break some standing with prometheus but might be acceptable.  So would become `agent_config_last_reload_timestamp_seconds` with label like `result=success` or `result=failure`.

File: cmd/agent/entrypoint.go
Comment (mattdurham) [inline cmd/agent/entrypoint.go:265]: Also in thinking about this I wonder if we need to change it to just `load`, or set values to good on startup. If I am alerting based on these they will not be valid til a reload happens.

File: pkg/config/config.go
Comment (mattdurham) [inline pkg/config/config.go:457]: nit: Robert likes comments to end with a .

File: cmd/agent/entrypoint.go
Comment (mattdurham) [inline cmd/agent/entrypoint.go:68]: Do these show up under the flow system? We should probably have them there to.

**Review by mattdurham — CHANGES_REQUESTED** [review]
Looks solid with some feedback requested.

File: cmd/agent/entrypoint.go
Comment (jcreixell) [reply cmd/agent/entrypoint.go:72]: I thought about, it's probably a good idea, let's do it

File: cmd/agent/entrypoint.go
Comment (gouthamve) [reply cmd/agent/entrypoint.go:265]: Shouldn't the default be success? This is because the "first" load of the config is a success and if its bad config, the startup itself will fail. I would default to success tbh.

Comment (jcreixell) [reply cmd/agent/entrypoint.go:265]: uhm, let's move this to config.go so that we keep track of loads and not just reloads

File: cmd/agent/flow_run.go
Comment (jcreixell) [inline cmd/agent/flow_run.go:39]: still not super happy about this one, as both flow and non-flow metrics show up in the metrics endpoint, instead of being registered lazily. Ideas welcome

**Comment by jcreixell** [comment]
@mattdurham I have addressed your comments, please have another look when you have time. The only remaining issue I can think of is that both flow and non-flow metrics are exposed, however I am not sure that fixing this is worth the hassle. Thank you!

File: cmd/agent/flow_run.go
Comment (tpaschalis) [reply cmd/agent/flow_run.go:39]: Do you mean that Flow metrics would show up even in the 'static' Agent mode?

If so, it might be a little annoying but not too much. We _could_ stuff everything into the Run method, but I'm not sure the readability is a good tradeoff.

**Review by tpaschalis — APPROVED** [review]
LGTM! Nothing to add personally, simple and straightforward.

File: cmd/agent/flow_run.go
Comment (mattdurham) [reply cmd/agent/flow_run.go:39]: For Flow side: In `flow_run.go` instead of using promauto you could register it with the DefaultRegister. Same with static agent , instead of promauto register with the reg in `entrypoint.go`

**Comment by jcreixell** [comment]
@mattdurham See my last commit for a refactoring moving the duplicated code into a new package. 

I tried the simple solution of manually registering the metrics during execution, but it was quite intrusive, especially in config.go (since Load can be called multiple times, I would need to register the metrics outside that package and pass them as a parameter). In the end, I opted for using a singleton instance for the metrics, reused both in flow and regular agent.

Location of this new package and naming are up for discussion.

File: pkg/config/instrumentation/config_metrics.go
Comment (mattdurham) [inline pkg/config/instrumentation/config_metrics.go:51]: The `isError` through me off a bit; I was expecting more success bool.

File: cmd/agent/flow_run.go
Comment (mattdurham) [inline cmd/agent/flow_run.go:112]: Could this be simplified to one `defer instrumentation.ConfigMetrics.InstrumentLoad(err != nil)`?

File: pkg/config/instrumentation/config_metrics.go
Comment (jcreixell) [reply pkg/config/instrumentation/config_metrics.go:51]: sounds good, let's change it

File: pkg/config/instrumentation/config_metrics.go
Comment (jcreixell) [inline pkg/config/instrumentation/config_metrics.go:31]: question: shall we be using the term `load` instead of `reload` here? Reload could be confusing given that we instrument every load, including the initial one

Comment (mattdurham) [reply pkg/config/instrumentation/config_metrics.go:31]: Sounds reasonable.

**Review by mattdurham — APPROVED** [review]
Looks good to me, clean and simple

File: pkg/config/instrumentation/config_metrics.go
Comment (rfratto) [inline pkg/config/instrumentation/config_metrics.go:37]: Hm, one of my concerns with the `result` label is that every instance of the label will continue to be reported for the process lifetime, which might be confusing and hard to query for exactly what you're looking for. 

I think there's two choices here:

1. Remove the result label 
2. Delete all the labels on the vector every time we instrument a new load, so the /metrics endpoint only reports the metric and label corresponding to the most recent load.  

Personally, I'd go for the first; a user can still know when unsuccessful loads happened by querying for `agent_config_last_load_timestamp_seconds and agent_config_last_load_successful == 0`.

File: pkg/config/instrumentation/config_metrics.go
Comment (rfratto) [inline pkg/config/instrumentation/config_metrics.go:51]: nit: Passing success to a function like this feels a little weird to me. You could implement this using some middleware-y function:

```go
cm.InstrumentLoad(bb, func(bb []byte) error {
  // do the load 
  // return if there was an error 
})
```

which would set all the appropriate metrics for you. 

Honestly, I'm not sure if that's really worth implementing right now. This is mostly just a nit, it feels a little strange but I wouldn't say it's a blocker.

**Review by rfratto — APPROVED** [review]
LGTM. I have some concerns about the result label for `agent_config_last_load_timestamp_seconds`, but how you solve it is ultimately up to you and you don't need a second approval from me. 

Let @grafana/grafana-agent-maintainers know when you're ready for it to be merged and someone can put it in main :)

File: pkg/config/instrumentation/config_metrics.go
Comment (rfratto) [reply pkg/config/instrumentation/config_metrics.go:37]: cc @mattdurham since I'm suggesting that the `result` label might cause issues

Comment (mattdurham) [reply pkg/config/instrumentation/config_metrics.go:37]: Seems reasonable.

Comment (rfratto) [reply pkg/config/instrumentation/config_metrics.go:37]: SGTM 👍

File: pkg/config/instrumentation/config_metrics.go
Comment (jcreixell) [reply pkg/config/instrumentation/config_metrics.go:55]: I agree with your point here, however looking at the config code, it is hard to agree on a single API that would work for all current implementations (including dynamic config). I am also purposely trying to separate config hash instrumentation from config load (under the assumption that it is better to build the hash before applying dynamic transformations like variable substitution, etc to prevent noise, while it is better to instrument loads as late as possible, to cover all possible failure modes). I am also open to have these assumptions challenged :)

Comment (rfratto) [reply pkg/config/instrumentation/config_metrics.go:55]: I don't feel too strongly about it, it's something we can always change later and it's not intended to be a public facing API. It's more than fine for now 👍

**Comment by jcreixell** [comment]
@grafana/grafana-agent-maintainers I think I am done here, feel free to merge if you are ok with my last commits and comment

**Pull request merged**
